site-src/geps/gep-713.md

```diff
@@ -0,0 +1,409 @@
+target more than one resource at a time, but it can be used target larger
+resources such as Gateways or Namespaces that may apply to multiple child
+resources. This field should have the following structure:
+
+```go
+// PolicyTargetReference identifies an API object to apply policy to.
+type PolicyTargetReference struct {
+    // Group is the group of the target resource.
+    //
+    // Support: Core
+    //
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=253
+    Group string `json:"group"`
+
+    // Kind is kind of the target resource.
+    //
+    // Support: Core
+    //
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=253
+    Kind string `json:"kind"`
+
+    // Name is the name of the target resource.
+    //
+    // Support: Core
+    //
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=253
+    Name string `json:"name"`
+
+    // SectionName is the name of a section within the target resource. When 
+    // unspecified, this targets the entire resource. In the following
+    // resources, SectionName is interpreted as the following:
+    // * Gateway: Listener Name
+    // * Route: Rule Name
+    // * Service: Port Name
+    //
+    // Support: Core
+    //
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=253
+    // +optional
+    SectionName string `json:"sectionName,omitempty"`
+
+    // Namespace is the namespace of the referent. When unspecified, the local
+    // namespace is inferred. Even when policy targets a resource in a different
+    // namespace, it may only apply to traffic originating from the same 
+    // namespace as the policy.
+    //
+    // Support: Extended
+    //
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=253
+    // +optional
+    Namespace string `json:"namespace,omitempty"`
+    
+    // ClassName is the name of the class this policy should apply to. When
+    // unspecified, the policy will apply to all classes that support it.
+    //
+    // Support: Core
+    //
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=253
+    // +optional
+    ClassName string `json:"className,omitempty"`
+}
+```
+
+### Status
+A new `Policies` field will be added to status on Gateways and Routes. This will
+be a list of `PolicyTargetReference` structs with the fields instead used to
+refer to the Policy resource that has been applied.
+
+### Hierarchy
+Policy precedence is determined in the same direction a request flows. For
+ingress, that means from Gateway to Backend, and for mesh that means from
+Consumer to Backend. 
+
+| Ingress | Sidecar Consumer |
+|-|-|
+| Gateway Namespace | Consumer Namespace |
+| Gateway | Consumer |
+| Route Namespace | Route Namespace |
+| Route | Route |
+| Backend Namespace | Backend Namespace |
+| Backend | Backend |
```

Comment:
Is this the configuration you're thinking of:

```
kind: ServicePolicy 
spec:
  default:
    requestTimeout: 30
  targetRef:
    kind: HTTPRoute
    name: my-route
---
kind: ServicePolicy 
spec:
  default:
    requestTimeout: 10
  targetRef:
    kind: Service
    name: reports-v1
```

But if `/v1/reports` maps to the `reports-v1` Service that takes longer to respond, then wouldn't the developer just want to increase the timeout for that Service directly then? It seems like in this case there's no need to use the HTTPRoute to customize the timeout for an individual path since the delay is caused by the Service behind that path anyways. Am I thinking about this correctly?

Reply:
For example, if I have /slow and /fast, and want different timeouts

Reply:
Oh you want to provide different timeouts to paths which are destined for the same Service? I think you would just need different routes for that since policies can't be applied within a resource. I'm not sure if it feels worth it to allow policies to reference at sections within a resource, at least not initially. The user can always create mirrored routes or services to apply per-routing rule policies.

```
kind: HTTPRoute
apiVersion: networking.x-k8s.io/v1alpha1
metadata:
  name: fast-timeout
spec:
  rules:
  - matches:
    - path:
        type: Prefix
        value: /fast
    forwardTo:
    - serviceName: my-service
      port: 8080
---
kind: ServicePolicy 
spec:
  default:
    requestTimeout: 10
  targetRef:
    kind: HTTPRoute
    name: fast-timeout
---
kind: HTTPRoute
apiVersion: networking.x-k8s.io/v1alpha1
metadata:
  name: slow-timeout
spec:
  rules:
  - matches:
    - path:
        type: Prefix
        value: /slow
    forwardTo:
    - serviceName: my-service
      port: 8080
---
kind: ServicePolicy 
spec:
  default:
    requestTimeout: 30
  targetRef:
    kind: HTTPRoute
    name: slow-timeout
```

Reply:
Yep, that's exactly it, I'm going to add a diagram like @mark-church suggested below that should make this a bit easier to understand.

Reply:
Missed the latest comment from @mark-church. Since we can attach policy to a `sectionName`, we'd be able to attach policy to a specific rule in a route, as long as that rule had a name. So instead we'd end up with this:


```yaml
kind: HTTPRoute
apiVersion: networking.x-k8s.io/v1alpha1
metadata:
  name: timeout-example
spec:
  rules:
  - name: fast 
    matches:
    - path:
        type: Prefix
        value: /fast
    forwardTo:
    - serviceName: my-service
      port: 8080
  - name: slow 
    matches:
    - path:
        type: Prefix
        value: /slow
    forwardTo:
    - serviceName: my-service
      port: 8080
---
kind: ServicePolicy 
spec:
  default:
    requestTimeout: 10
  targetRef:
    kind: HTTPRoute
    name: timeout-example
    sectionName: fast
---
kind: ServicePolicy 
spec:
  default:
    requestTimeout: 30
  targetRef:
    kind: HTTPRoute
    name: timeout-example
    sectionName: slow
```

Reply:
The solution above solutions work only if there is no ServicePolicy associated with `my-service`. What if one wishes to override the timeout in one route _and_ have a default at the Service?

Would something like the following be supported?


```yaml
kind: ServicePolicy 
spec:
  enforce: # NOTE THIS
    requestTimeout: 30
  targetRef:
    kind: HTTPRoute
    name: my-slow-route
---
kind: ServicePolicy 
spec:
  default:
    requestTimeout: 10
  targetRef:
    kind: Service
    name: reports-v1
```

Here, the default request-timeout is 10seconds and is configured at the Service-level for all requests, no matter which path the request traverses with one exception: if the request traverses my-slow-route, then the request-timeout will be 30 seconds.

Reply:
Yep, that's a good example the motivation for the override/enforce field.